README
NAME
    pwmarker - generate pinwin-style markers with perspective shadows.

SYNOPSIS
     width = 240
     height = 180
     anchor = 50
                
     m = PinwinMarker(width, height, anchor)
     m.draw()
 
     m.save("/path/to/marker.png")

DESCRIPTION
    pwmarker is a class for generating pinwin-style markers with either a
    stylized ("cartoon") or a true perspective shadow. Pinwin markers are
    designed to act as containers for other images, or text, but adding
    either is outside the scope of this package and left up to individual
    developers.

    Markers are returned as PNG images with a transparent background and
    semi-transparent shadows.

    By default pwmarker tries to use the Python Cairo bindings as a
    rendering engine but if they are not installed then it will use the
    standard Python Image Library (PIL). In either case PIL is required.

PACKAGE METHODS
  PinwinMarker(width, height, anchor=0)
    *width* and *height* are the dimensions of the source image that you are
    creating a pinwin for. Additional padding is added programatically
    relative to the size of the image no more or no less than 25 and 15
    pixels respectively.

    *anchor* is the height of pinwin's anchor. If no value is defined then
    it will be calculated as one-third of the source image's height but no
    less than 20 pixels.

    Returns a *PinwinMarker* object!

OBJECT METHODS
  obj.draw()
    This method renders the multiple layers for the pinwin, each with its
    own transparent background :

    * pinwin
        Just the pinwin.

    * shadow
        Just the shadow.

    * all (or combined)
        The pinwin marker overlayed on the shadow layer.

    Returns null.

  obj.save(path, layer='all')
    Write a specific marker layer to a file. Valid layers are :

    * pinwin
        Just the pinwin.

    * shadow
        Just the shadow.

    * all (default)
        The pinwin marker overlayed on the shadow layer.

    Returns true or false.

MARKER PROPERTIES YOU SHOULD CARE ABOUT
    Once you've created a marker you will probably want to place an image on
    the pinwin's "canvas" and place the marker itself on another image.

    The following object properties will be set after you've called the
    *draw* method :

    * padding
        The number of pixels, on both the x and y axis, offset from (0, 0)
        to center an image on the pinwin "canvas".

    B *<pt_x>
        The number of pixels, on the x axis, offset from (0, 0) representing
        the bottom (or point) of the pinwin's anchor.

    B *<pt_y>
        The number of pixels, on the y axis, offset from (0, 0) representing
        the bottom (or point) of the pinwin's anchor.

MARKER PROPERTIES YOU MAY CARE ABOUT
    The following object properties can be tweaked before calling the *draw*
    method.

    * border
        Int.

        Set the border width for the pinwin. (This only works if you have
        the Python Cairo bindings installed.)

        Default is *2*

    * cartoon_shadow
        Boolean.

        Toggle between stylized "cartoon" shadows and shadows generated
        using a proper 45 degree perpective transformation.

        Default is *True*

    * blurry_shadows
        Boolean.

        Toggle whether shadows should be blurred and softened.

        Default is *True*

    * add_dot
        Boolean.

        Toggle whether to include a (pink) visual marker at the base of a
        pinwin's anchor.

        Default is *True*

VERSION
    1.0

DATE
    $Date: 2008/11/30 03:44:04 $

AUTHOR
    Aaron Straup Cope

EXAMPLE
    http://flickr.com/photos/straup/2599411504/

SEE ALSO
    <http://www.pythonware.com/products/pil/>

    <http://www.cairographics.org/pycairo/>

    <http://modestmaps.com/examples-python-ws/>

COPYRIGHT
    Copyright (c) 2008 Aaron Straup Cope. All Rights Reserved.

LICENSE
    http://www.modestmaps.com/license.txt



Changes
# -*-n3-*-

@prefix dc:      <http://purl.org/dc/elements/1.1/> .
@prefix dcterms: <http://purl.org/dc/terms/> .
@prefix doap:    <http://usefulinc.com/ns/doap#> .
@prefix asc:     <http://aaronland.info/ns/changefile/> .
                                     
<>
        dc:identifier           "$Id: Changes,v 1.2 2008/11/30 03:44:04 asc Exp $" ;
        dc:description          "Python classes for creating pinwin-style markers" ;
        dcterms:references      <http://aaronland.info/python/pwmarker/> .

<http://aaronland.info/python/pwmarker/pwmarker-1.0.tar.gz>
        dcterms:isVersionOf <http://aaronland.info/python/pwmarker/> ;

        doap:Version    [
                        doap:revision     "1.0" ;
                        doap:created      "2008-07-24" ;
                        ];
        asc:changes     [
                        asc:commit        "Initial public release." ;
                        ] .

<http://aaronland.info/python/pwmarker/>
        a doap:Project ;
        doap:download-page      "http://aaronland.info/python/pwmarker/" ;
        doap:download-mirror    "" .


pwmarker/text.py
# THIS SHOULD BE CONSIDERED DEPRECATED.

import cairo
import re

class text :

    def __init__ (self) :

        self.surface = None

        self.font_face_header = "Helvetica Bold"
        self.font_size_header = 18
        
        self.font_face_body = "Helvetica Bold"
        self.font_size_body = 11

        self.width = 0
        self.height = 0
        
        self.max_width = 0
        
        self.margin = 20
        self.ln_height_body = 0

    #
    
    def draw (self, header, body) :

        (header, body) = self.calculate(header, body)
        
        self.surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, self.width, self.height)
        ctx = cairo.Context(self.surface)

        # basic setup
        
        ctx.move_to(0, 0)
        ctx.rectangle(0, 0, self.width, self.height)
        ctx.set_source_rgb(255, 255, 255)
        ctx.fill()

        ctx.set_source_rgb(0, 0, 0)

        # first the header
        
        x = 0 + self.margin
        y = 0 + (self.margin * 2)
        ctx.move_to(x, y)

        ctx.select_font_face(self.font_face_header, cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_BOLD)    
        ctx.set_font_size(self.font_size_header)

        for part in header :
            ctx.show_text(part)
            y = y + int(self.margin * 1.5)
            ctx.move_to(x, y)

        # now the body
        
        ctx.select_font_face(self.font_face_body, cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_BOLD)    
        ctx.set_font_size(self.font_size_body)

        # fix me : allow for multiple paragraphs
        
        for txt in body :
            ctx.show_text(txt)
            y = y + self.ln_height_body
            ctx.move_to(x, y)       

    #
    
    def calculate (self, header, body) :

        chunks_header = []
        chunks_body = []
        
        self.surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, self.width, self.height)
        ctx = cairo.Context(self.surface)

        # fix me : chunkify the header
        
        ctx.select_font_face(self.font_face_header, cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_BOLD)
        ctx.set_font_size(self.font_size_header)

        (tw, th) = self.measure_text(ctx, header)

        if tw > (self.width - (self.margin * 2)) :
            self.width = tw + (self.margin * 2)

        chunks_header = self.chunkify(ctx, header)

        #
        
        ctx.select_font_face(self.font_face_body, cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_BOLD)    
        ctx.set_font_size(self.font_size_body)

        chunks_body = self.chunkify(ctx, paras)

        # sort out line height

        self.ln_height_body = int(max(map(lambda word: word[1], map(lambda ln: self.measure_text(ctx, ln), chunks_body))) * 1.5)

        #
        
        self.height += int(self.margin * (3 + len(chunks_body)))

        self.width = int(self.width)
        self.height = int(self.height)

        return (chunks_header, chunks_body)

    #

    def chunkify_header(self, ctx, header) :
        return self.chunkify(ctx, header)

    #
    
    def chunkify_body (self, ctx, paras) :

        chunks = []

        for txt in paras :
            chunks.extend(self.chunkify(ctx, txt))

        return chunks

    #
    
    def chunkify (self, ctx, txt) :

        chunks = []
        chunk = txt

        okw = self.width - (self.margin * 2)
    
        while len(chunk) :
            (dw, dh) = self.measure_text(ctx, chunk)

            self.height += dh
        
            if dw < okw :
                chunks.append(chunk)
                break

            foo = ''
        
            for i in range(0, int(dw)) :
                foo += chunk[i]

                (fw, fh) = self.measure_text(ctx, foo)

                if fw < okw :
                    continue
                else :
                    stuff = foo[:-1]

                match = re.findall("(\s([^\s]+))$", stuff)

                if len(match) :
                    rest = match[0][0]
                    l = len(rest)
                    stuff = stuff[:-l]
                    
                    chunks.append(stuff)
                    chunk = rest.strip() + chunk[i:]
                    
                    break

        return chunks
        
    #
    
    def save (self, path) :
        self.surface.write_to_png(path)

    #

    def measure_text (self, ctx, txt):
            
        xbearing, ybearing, width, height, xadvance, yadvance = (
            ctx.text_extents(unicode(txt)))
        
        return (width, height)

    #
    
if __name__ == '__main__' :

    title = "this is the header"
    paras = ["this is the body. it is very long, and has many run on sentences"]

    t = text()
    t.draw(title, paras)

    t.save("/home/asc/Desktop/t.png")


pwmarker/pwpil.py
__package__    = "pwmarker/pwpil.py"
__version__    = "1.0.1"
__author__     = "Aaron Straup Cope"
__url__        = "http://www.aaronland.info/python/pwmarker"
__date__       = "$Date: 2008/11/30 03:44:04 $"
__copyright__  = "Copyright (c) 2008 Aaron Straup Cope. All rights reserved."
__license__    = "http://www.modestmaps.com/license.txt"

import math

import PIL.Image
import PIL.ImageDraw
import PIL.ImageFilter

class PILMarker :

    #
    
    def p__dot (self, ctx='pinwin', dr=None) :
    
        # dr should never be None, it's just a
        # hack to make python stop whinging...
        
        x = self.pt_x + 1
        y = self.pt_y + 1
        
        x1 = x - self.dot_r
        y1 = y - self.dot_r
        x2 = x + self.dot_r
        y2 = y + self.dot_r

        if ctx.startswith('mask-') :
            sh_c = (255, 255, 255)
            dot_c = (255, 255, 255)
        else :
            sh_c = (42, 42, 42)            
            dot_c = (255, 0, 132)

	# mock shadow
        dr.ellipse((x1 + 1 , y1 + 2, x2 + 1, y2 + 2), fill=sh_c)        

	# the dot
        dr.ellipse((x1, y1, x2, y2), fill=dot_c)

        # no donut hole for PIL

    # 
    
    def p__pinwin(self, anchor='bottom', dot_ctx='pinwin', fill='white') :

        (w, h) = self.calculate_dimensions(anchor)

        # to prevent the de-facto border (artifacting)
        # from being cropped when a pinwin is rendered
        # without its shadow
        
        w += 1
        
        im = PIL.Image.new('RGBA', (w, h))
        
        coords = self.p__coords()
        return self.p__draw_pinwin(im, coords, dot_ctx, fill)
        
    #

    def p__draw_pinwin (self, im, coords, dot_ctx='pinwin', fill='white') :

        dr = PIL.ImageDraw.Draw(im)
        
        if self.add_dot and (dot_ctx != 'shadow' and dot_ctx != 'mask-shadow') :
            self.dot(dot_ctx, dr)

        dr.polygon(coords, fill=fill)

        # top left
        
        x1 = self.offset
        y1 = self.offset
        x2 = self.offset + self.padding * 2
        y2 = self.offset + self.padding * 2
        
        dr.pieslice((x1, y1, x2, y2), 180, 270, fill=fill)

        # bottom left
        
        x1 = self.offset
        y1 = self.offset + self.img_h 
        x2 = self.offset + self.padding * 2
        y2 = self.offset + (self.padding * 2) + self.img_h 
        
        dr.pieslice((x1, y1, x2, y2), 90, 180, fill=fill)

        # top right
        
        x1 = self.offset + self.img_w
        y1 = self.offset
        x2 = self.offset + self.img_w + (self.padding * 2)
        y2 = self.offset + (self.padding * 2)

        dr.pieslice((x1, y1, x2, y2), 270, 0, fill=fill)

        # bottom right
        
        x1 = self.offset + self.img_w
        y1 = self.offset + self.img_h
        x2 = self.offset + self.img_w + (self.padding * 2)
        y2 = self.offset + self.img_h + (self.padding * 2)

        dr.pieslice((x1, y1, x2, y2), 0, 90, fill=fill)
        return im
        
    #
    
    def p__shadow (self, anchor, dot_ctx, fill='black') :
        sh = self.p__pinwin(anchor, dot_ctx, fill)
        return self.tilt(sh, self.blurry_shadows)

    #
    
    def p__cartoon_shadow (self, anchor, dot_ctx, fill='black') :

        # make the canvas

        w = self.offset + self.img_w + (self.padding * 2)
        h = self.offset + self.img_h + (self.padding * 2)
        
        cnv = PIL.Image.new('RGBA', (w, h))
        coords = self.p__cartoon_shadow_coords()

        blur = False
        
        cnv = self.p__draw_pinwin(cnv, coords, dot_ctx, fill)
        cnv = self.tilt(cnv, blur)    
        
        coords = self.calculate_cartoon_anchor_coords(cnv)

        (w, h, sh_offset) = coords[0]
        (sha_left, cnv_h) = coords[1]
        (bottom_x, bottom_y) = coords[2]
        (sha_right, cnv_h) = coords[3]
                
        sh = PIL.Image.new('RGBA', (w, h))
        sh.paste(cnv, (sh_offset, 0), cnv)
        
        dr = PIL.ImageDraw.Draw(sh)
                
        anchor = [
            (sha_left, cnv_h),
            (bottom_x, bottom_y),        
            (sha_right, cnv_h),
            (sha_left, cnv_h),            
            ]

        dr.polygon(anchor, fill=fill)
        
        # dot

        self.dot('shadow', dr)
    
        # blur!

        if not self.blurry_shadows :
            return sh
        
        return self.p__blur(sh)
    
    #

    def p__coords (self) :

        #
        #    startx, starty
        #
        #      nwa ------------- nea
        #      /                   \ 
        #    nwb                   neb   
        #    |                       |
        #    |                       |
        #    swb                   seb
        #      \   cna   cnc       /
        #      swa --/ | /------ sea
        #             |
        #            |
        #           |
        #          |
        #         cnb
        #
        #                  endx, endy
        #
        # The rounded corners get bolted on after the
        # fact - this is dumb and if I can ever find
        # docs/examples for PIL's ImagePath stuff then
        # I will use that instead...
        
        nwa_x = self.offset + self.padding
        nwa_y = self.offset

        nea_x = self.offset + self.padding + self.img_w
        nea_y = self.offset

        neb_x = self.offset + self.img_w + (self.padding * 2)
        neb_y = self.offset + self.padding

        seb_x = self.offset + self.img_w + (self.padding * 2)
        seb_y = self.offset + self.img_h + self.padding

        sea_x = self.offset + self.img_w + self.padding
        sea_y = self.offset + self.img_h + (self.padding * 2)

        cnc_x = self.offset + self.offset_cone + int(self.anchor_w * .5)
        cnc_y = self.offset + self.img_h + (self.padding * 2)

        cnb_x = self.offset + self.offset_cone
        cnb_y = self.offset + self.canvas_h

        cna_x = self.offset + self.offset_cone - int(self.anchor_w * .5)
        cna_y = self.offset + self.img_h + (self.padding * 2)
        
        swa_x = self.offset + self.padding
        swa_y = self.offset + self.img_h + (self.padding * 2)

        swb_x = self.offset
        swb_y = self.offset + self.img_h + self.padding

        nwb_x = self.offset
        nwb_y = self.offset + self.padding

        frame = [(nwa_x, nwa_y),
                  (nea_x, nea_y), (neb_x, neb_y),
                  (seb_x, seb_y), (sea_x, sea_y),
                  (cnc_x, cnc_y), (cnb_x, cnb_y),
                  (cna_x, cna_y), (swa_x, swa_y),
                  (swb_x, swb_y), (nwb_x, nwb_y),
                  (nwa_x, nwa_y)]

        return frame

    #
    
    def p__cartoon_shadow_coords (self) :
        
        nwa_x = self.offset + self.padding
        nwa_y = self.offset

        nea_x = self.offset + self.padding + self.img_w
        nea_y = self.offset

        neb_x = self.offset + self.img_w + (self.padding * 2)
        neb_y = self.offset + self.padding

        seb_x = self.offset + self.img_w + (self.padding * 2)
        seb_y = self.offset + self.img_h + self.padding

        sea_x = self.offset + self.img_w + self.padding
        sea_y = self.offset + self.img_h + (self.padding * 2)
        
        swa_x = self.offset + self.padding
        swa_y = self.offset + self.img_h + (self.padding * 2)

        swb_x = self.offset
        swb_y = self.offset + self.img_h + self.padding

        nwb_x = self.offset
        nwb_y = self.offset + self.padding

        frame = [(nwa_x, nwa_y),
                 (nea_x, nea_y),
                 (neb_x, neb_y),
                 (seb_x, seb_y),
                 (sea_x, sea_y),
                 (swa_x, swa_y),                 
                 (swb_x, swb_y),
                 (nwb_x, nwb_y),
                 (nwa_x, nwa_y)]

        return frame
    
    #
    
    def p__tilt(self, pil_im, blur=True) :

        # foo is to remind me that math is hard...
        
        foo = 3.75
        
        (iw, ih) = pil_im.size

        iw2 = int(float(ih) * .45) + iw
        ih2 = int(ih / foo)

        bar = int(float(ih2) * 1.6)

        a = 1
        b = math.tan(45)
        c = -bar
        d = 0
        e = foo
        f = 0
        g = 0
        h = 0
        
        data = (a, b, c, d, e, f, g, h)

        return pil_im.transform ((iw2,ih2), PIL.Image.PERSPECTIVE, data, PIL.Image.BILINEAR)

    #
    
    def p__blur(self, pil_im, iterations=5) :

        (w, h) = pil_im.size

        # ensure some empty padding so the blurring
        # doesn't create butt ugly lines at the top
        # and bottom
        
        h+= 20
        
        im = PIL.Image.new('RGBA', (w, h))
        dr = PIL.ImageDraw.Draw(im)

        im.paste(pil_im, (0, 10), pil_im)
        
        for i in range(1, iterations) :
            im = im.filter(PIL.ImageFilter.BLUR)

        return im    

    #

    def p__antialias(self, im) :

        scale = 4
        sz = im.size
        im = im.resize((sz[0] * scale, sz[1] * scale),  PIL.Image.NEAREST)
        im = im.resize(sz, PIL.Image.ANTIALIAS)

        return im


pwmarker/pwcommon.py
__package__    = "pwmarker/pwcommon.py"
__version__    = "1.0.1"
__author__     = "Aaron Straup Cope"
__url__        = "http://www.aaronland.info/python/pwmarker"
__date__       = "$Date: 2008/11/30 03:44:04 $"
__copyright__  = "Copyright (c) 2008 Aaron Straup Cope. All rights reserved."
__license__    = "http://www.modestmaps.com/license.txt"

import math

import PIL.Image
import PIL.ImageDraw

class Common :

    def draw (self, anchor='bottom') :

        pw = self.generate_pinwin(anchor)
        sh = self.generate_shadow(anchor)

        sh = self.position_shadow(pw, sh)
        
        for ctx in self.rendered.keys() :
            mask = self.generate_mask(anchor, ctx)
        
            if ctx == 'pinwin' :
                im = self.render_pinwin(pw, mask)
            elif ctx == 'shadow' :
                im = self.render_shadow(pw, sh, mask)        
            else :
                im = self.render_all(pw, sh, mask)   

            if self.render_engine == 'pil' :
                im = self.p__antialias(im)
                
            self.rendered[ctx] = im

    # 
    
    def generate_pinwin (self, anchor='bottom', dot_ctx='pinwin', *args) :

        key = "%s-%s-%s" % (anchor, dot_ctx, str(args))

        if self.pinwin_cache.has_key(key) :
            return self.pinwin_cache[key]

        #
        
        if self.render_engine == 'cairo' :

            pw = self.c__pinwin(anchor, dot_ctx, *args)       
            pw = self.c__cairo2pil(pw)        
        else :
            pw = self.p__pinwin(anchor, dot_ctx, *args)

        #
        
        self.pinwin_cache[key] = pw
        return pw
        
    #
    
    def generate_shadow (self, anchor='bottom', dot_ctx='shadow', *args) :

        # not clear whether caching is actually
        # necessary or useful here ...
        
        key = "%s-%s-%s" % (anchor, dot_ctx, str(args))

        if self.cartoon_shadows :
            key += "-cartoon"

        if self.shadow_cache.has_key(key) :
            return self.shadow_cache[key]

        #
        
        if self.render_engine == 'cairo' :

            if self.cartoon_shadows :
                sh = self.c__cartoon_shadow(anchor, dot_ctx, *args)
            else :
                sh = self.c__shadow(anchor, dot_ctx, *args)
                sh = self.c__cairo2pil(sh)

        else :
            if self.cartoon_shadows :
                sh = self.p__cartoon_shadow(anchor, dot_ctx, *args)
            else :
                sh = self.p__shadow(anchor, dot_ctx, *args)            

        #
        
        self.shadow_cache[key] = sh
	return sh

    #

    def generate_mask (self, anchor='bottom', ctx='all') :

        dot_ctx = "mask-%s" % ctx

        pw_c = (255, 255, 255)

        if self.render_engine == 'cairo' :
            sh_c = (.65, .65, .65)
        else :
            sh_c = (130, 130, 130)
        
        pw = self.generate_pinwin(anchor, dot_ctx, pw_c)
        sh = self.generate_shadow(anchor, dot_ctx, sh_c)
        
        (pww, pwh) = pw.size
        (shw, shh) = sh.size
        
        w = max(pww, shw)
        h = max(pwh, shh)

        if ctx == 'pinwin' :
            w = pww
            h = pwh

        im = PIL.Image.new('L', (w, h))
        dr = PIL.ImageDraw.Draw(im)

        if ctx != 'pinwin' : 
            shx = 3
            shy = h - shh

            # ugly ugly hack...please to track down
            # how not to need to do this...
            
            if self.add_dot :
                if self.cartoon_shadows :
                    shy -= int(self.dot_r * 1)
                else :
                    shy -= int(self.dot_r * 1.5)
                    shx -= int(self.dot_r * .5)
                    
            im.paste(sh, (shx, shy), sh)
            
        if ctx != 'shadow' :
            im.paste(pw, (1, 1), pw)
                
        return im

    #

    def tilt(self, pil_im, blur=True):

        sh = self.p__tilt(pil_im, blur)

        if blur == True :
            sh = self.blur(sh)

        return sh

    #
    
    def blur(self, pil_im) :
        return self.p__blur(pil_im)
    
    #

    def dot (self, ctx='pinwin', *args) :

        if self.render_engine == 'cairo' :
            return self.c__dot(ctx, args)
        else :
            return self.p__dot(ctx, *args)
        
    #
    
    def calculate_dimensions (self, anchor='bottom', ctx='pinwin') :

        if not self.border_c :
            self.border_c = (0, 0, 0)
            
        if not self.border_w :
            self.border_w = 2

        self.offset = int(math.ceil(self.border_w / 2))
        
        self.padding = int(min(self.img_w, self.img_h) * .1)

        if self.padding < 15 :
            self.padding = 15
        elif self.padding > 25 : 
            self.padding = 20
        else :
            pass

        self.corner_r = self.padding
        
        self.canvas_w = self.offset + self.border_w + self.img_w + (self.padding * 2)
        self.canvas_h = self.offset + self.img_h + (self.padding * 2) + self.anchor_h

        self.offset_x = self.offset + self.padding
        self.offset_y = self.offset + self.padding
            
        self.offset_cone = int(self.canvas_w * .35)

        if self.anchor_w > int(self.canvas_w / 2) :
            self.anchor_w = int(self.canvas_w / 3)

        h = self.canvas_h

        if ctx == 'pinwin' and self.add_dot :
            h += int(self.dot_r * 2)
            
        self.pt_x = self.offset + self.offset_cone
        self.pt_y = self.offset + self.img_h + (self.padding * 2) + self.anchor_h

        # legacy crap (see above)
        # note the +1 ... not sure, but it's necessary
        
        self.x_padding = self.offset + self.padding + 1
        self.y_padding = self.offset + self.padding + 1
        self.x_offset = self.pt_x
        self.y_offset = self.pt_y

        return (self.canvas_w, h)

    #

    def calculate_cartoon_anchor_coords (self, tilted_cnv) :

        (cnv_w,cnv_h) = tilted_cnv.size

        key = "%s-%s" % (cnv_w, cnv_h)

        if self.cartoon_anchor_cache.has_key(key) :
            return self.cartoon_anchor_cache[key]
        
        #
        
        w = self.offset + self.img_w + (self.padding * 2)
        h = self.offset + self.img_h + (self.padding * 2)

        sh_offset = int(w * .2)
        sh_anchor_h = int(self.anchor_h * .9)

        sh_anchor_w = int(self.anchor_w * .9)
        
        diff = self.anchor_h - sh_anchor_h
        cnv_half = int(cnv_h / 2)

        # for pinwins with insanely tall anchors
        # 90% may be larger than the canvas' height
        # which just looks weird

        if sh_anchor_h > cnv_h :
            sh_anchor_h = self.anchor_h - cnv_half

        # but we also want to make sure that shadows
        # for pinwins with short stubby anchors don't
        # start flush with the bottom of the canvas...
        # these are "cartoon" shadows after all so a
        # certain amount of artistic license is allowed
        
        elif diff < int(cnv_h / 4) :

            pw_cnv_h = (self.img_h + (self.padding * 2))
            
            if self.anchor_h < int(pw_cnv_h / 2) :
                sh_anchor_h = int(self.anchor_h / 2)
            else :
                sh_anchor_h = self.anchor_h - int(cnv_h / 4)
            
        else :
            pass
            
        # set up the dimensions for the image we're going to
        # combine the tilted canvas and anchor on
        
        tmp_w = cnv_w + sh_offset
        tmp_h = cnv_h + sh_anchor_h

        # figure out the coordinates for the anchor
        
        pwa_left = (self.offset + self.offset_cone - int(self.anchor_w * .5))
        pwa_right = pwa_left + self.anchor_w
        
        sha_left = pwa_left + sh_offset
        sha_right = sha_left + sh_anchor_w
        
        bottom_x = self.offset + self.offset_cone - self.border_w
        bottom_y = tmp_h

        # we don't draw the do because we'll never see it
        # but we do need to consider it

        if self.add_dot :
            bottom_y = bottom_y - self.dot_r

        if self.blurry_shadows :
            bottom_y += int(self.dot_r * 2.5)
            
        coords = ((tmp_w, tmp_h, sh_offset),
                  (sha_left, cnv_h),
                  (bottom_x, bottom_y),
                  (sha_right, cnv_h))
        
        self.cartoon_anchor_cache[key] = coords
        return coords
    
    #
    
    def position_shadow (self, pw, sh) :

        (pww, pwh) = pw.size
        (shw, shh) = sh.size

        w = max(pww, shw)
        h = max(pwh, shh)

        im = PIL.Image.new('RGBA', (w, h))
        dr = PIL.ImageDraw.Draw(im)

        shx = 3
        shy = h - shh

        if self.add_dot :
            shy -= int(self.dot_r)
        
        im.paste(sh, (shx, shy), sh)
        return im
    
    #

    def render_all (self, pw, sh, mask) :

        im = self.position_shadow(pw, sh)
        im.paste(pw, (1, 1), pw)
        
        im.putalpha(mask)
    	return im

    #
    
    def render_pinwin (self, pw, mask) :

	(w, h) = pw.size
        
        im = PIL.Image.new('RGBA', (w, h))
        dr = PIL.ImageDraw.Draw(im)

        im.paste(pw, (1, 1), pw)
        im.putalpha(mask)
    	return im

    #

    def render_shadow (self, pw, sh, mask) :

        im = self.position_shadow(pw, sh)
        im.putalpha(mask)
    	return im
            
    #

    def save (self, path, ctx='all') :
        if not self.rendered.has_key(ctx) :
            return None
        
        self.rendered[ctx].save(path)

    # backwards compatibility with (ws-modestmaps/markers.py)
    
    def fh (self, ctx='all') :
        if not self.rendered.has_key(ctx) :
            return None
        
        return self.rendered[ctx]

    # 
